modules/core/src/action.ts

```diff
@@ -1,27 +1,28 @@
 import {curry2} from 'ts-curry'
+
 import {isNil} from './isNil'
+import {INilAction} from './nil'
 
 /**
  * Interface for Action
- * @interface
  */
-export interface Action<T> {
+export interface IAction<T> {
   type: string | number
   value: T
 }
 
-function createAction<T>(type: string | number, value: T) {
+function createAction(type: string | number, value: INilAction): INilAction
+function createAction<T>(type: string | number, value: T): IAction<T>
+function createAction(type: string | number, value: unknown): IAction<unknown> {
   return isNil(value) ? value : {type, value}
 }
 
 /**
  * Creates a new Action type object
- * @function
- * @param {string|number} type
- * @param {any} value
- * @returns {Action}
+ * @param: Action Type
+ * @param: Action Payload
  */
 export const action: {
-  <T>(type: string | number, value: T): Action<T>
-  <T>(type: string | number): {(value: T): Action<T>}
+  <T>(type: string | number, value: T): IAction<T>
+  <T>(type: string | number): (value: T) => IAction<T>
```

Comment:
even though `createAction` has good typings now, the inferred return type is overwritten here.
```typescript
const actual = action('WIND', Nil()) // IAction<INilAction>
```

Ref: https://github.com/action-land/action-land/pull/29/files#diff-ade9c5c4a4ecd2bb9c3785ac934117e2R26